build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -160,6 +160,12 @@ service ActionCache {
   // [Command][build.bazel.remote.execution.v2.Command], into the
   // `ContentAddressableStorage`.
   //
+  // Server implementations are advised to modify the uploaded ActionResult to
```

Comment:
I do struggle with encouraging modification of ActionResult.

Reply:
It's implicit that ActionResults can be modified, otherwise why does UpdateActionResult return an ActionResult?

Would it help if we suggest execution_metadata.worker here too, as a safe and convenient value to modify?

Reply:
> It's implicit that ActionResults can be modified, otherwise why does UpdateActionResult return an ActionResult?

Actually this raises a good point.  To me it isn't implicit at all that ActionResults can be modified.  It could be returning the ActionResult that was replaced by the new ActionResult.

Regarding modification, I would rather that the ActionCache service is explicitly not allowed to modify ActionResult.  When when we start considering signing of ActionResult, we won't have to worry about ActionCache.

Reply:
Updated the text, so that it's simpler and softer, but still hopefully useful to server implementors.

I was also confused by the fact that `UpdateActionResult` returns an `ActionResult`, then I stumbled upon this (note the discussion about mutable resource properties): https://cloud.google.com/apis/design/standard_methods#update

Reply:
> Regarding modification, I would rather that the ActionCache service is explicitly not allowed to modify ActionResult. When when we start considering signing of ActionResult, we won't have to worry about ActionCache.

If you mean workers signing what they did, I don't think signing ActionResult as-is is the way to go. Some fields - like stdout_raw, execution_metadata.worker, execution_metadata.queued_timestamp - are explicitly intended for servers to set, either in general or conditional on the cache hit request. And that will likely expand, as ActionResult is a communication message used directly in RPC responses. But more concerningly, ActionResult doesn't describe the action itself, and any use-case for worker signing where the server is untrusted presumably requires the worker sign _what they ran_, not just the result of it, otherwise you're subject to replay attacks of various sorts.

Whatever workers sign should be

a.    self-contained: no extraneous fields, no optional fields introducing ambiguity, well-defined serialization, etc
b.    complete: cover everything material to execution, including inputs, outputs, command, environment changes, etc
c.    passed through the API as-is without modification alongside the corresponding signature.

I'm not sure how exactly to achieve this, especially with the additional complication that REAPI isn't even a worker-facing API (the APIs in use to communicate with workers may not reuse any messages from it - RWAPI reuses the CAS but none of Execution, for example - but regardless, I'm pretty sure signing ActionResult as-is is not going to be the way to go, so it being flagged as server-modifiable is probably useful in not misleading anyone in that direction.

Reply:
@EricBurnett I don't think we are disagreeing on signing ActionResult as-is, and yes, at a minimum the Action digest should be included in the ActionResult.  This would also solve the defaults-only ActionResult problem.  We can take the ActionResult signing discussion to a separate issue (eg. #133) though :)

To bring it back to this issue, I would prefer to keep ActionCache as a simple store/retrieve service.

I would be fine with a description along the lines of:
"The server MAY return an equivalent ActionResult as a response.  The client SHOULD use this equivalent ActionResult for inputs to dependent Actions remaining to be executed."

This does not imply that the ActionCache service munges ActionResults on the way in.  It does imply that ActionCache may return a "better" result for the client to use moving forward.  Now, how this "better" result comes to be is an implementation detail.  It may be selective based on which process produced the result for the action, or when, or...

What I like about this phrasing is that there is something useful for the client in the response as well.

Hope that clarifies my thinking a little bit.

Reply:
> at a minimum the Action digest should be included in the ActionResult. This would also solve the defaults-only ActionResult problem

Many Actions can produce the same ActionResult, so it would have to be a repeated field I guess.

> I would be fine with a description along the lines of:
> "The server MAY return an equivalent ActionResult as a response. The client SHOULD use this equivalent ActionResult for inputs to dependent Actions remaining to be executed."

This part sounds fine to me. I would also like to hint that zero-valued ActionResults should be avoided (though I have not yet found another cache implementor willing to comment here).

Reply:
@mostynb I think that language allows for server implementations to not store zero-valued ActionResults as-is.

Would you like to add something like:
- In UpdateActionResult "Clients SHOULD not send zero-sized ActionResults"
- In GetActionResult "Servers SHOULD not return zero-sized ActionResults"

Or do you have something completely different in mind?

Reply:
Currently on master, there is no mention of the UpdateActionResult return value here. So I assume that the "Server implementations MAY modify the ActionResult and return an equivalent value" part is uncontroversial.

The second part of this PR is merely a hint for server implementors as to why they might want to do so- as a way to catch the most common form of data corruption while the ActionResult is in storage (and ActionResult.execution_metadata.Worker is the simplest field to use for this purpose). How about something like:

`Servers MAY modify fields such as ActionResult.execution_metadata.Worker, to ensure that  the serialized value of the ActionResult is non-empty during processing and storage.`

But even "MAY" feels too strong IMO. Would you prefer that I drop this part entirely?